lib/client/index.js

```diff
@@ -0,0 +1,1710 @@
+      updateBrushToNow();
+    }
+    updateChart(false);
+  }
+
+  function visibilityChanged() {
+    var prevHidden = documentHidden;
+    documentHidden = (document.hidden || document.webkitHidden || document.mozHidden || document.msHidden);
+
+    if (prevHidden && !documentHidden) {
+      console.info('Document now visible, updating - ' + (new Date()));
+      refreshChart(true);
+    }
+  }
+
+  window.onresize = refreshChart;
+
+  document.addEventListener('webkitvisibilitychange', visibilityChanged);
+
+
+  updateClock();
+
+  function Dropdown(el) {
+    this.ddmenuitem = 0;
+
+    this.$el = $(el);
+    var that = this;
+
+    $(document).click(function() { that.close(); });
+  }
+  Dropdown.prototype.close = function () {
+    if (this.ddmenuitem) {
+      this.ddmenuitem.css('visibility', 'hidden');
+      this.ddmenuitem = 0;
+    }
+  };
+  Dropdown.prototype.open = function (e) {
+    this.close();
+    this.ddmenuitem = $(this.$el).css('visibility', 'visible');
+    e.stopPropagation();
+  };
+
+  var silenceDropdown = new Dropdown('.dropdown-menu');
+
+  $('.bgButton').click(function (e) {
+    if (alarmingNow()) {
+      silenceDropdown.open(e);
+    }
+  });
+
+  $('#silenceBtn').find('a').click(function (e) {
+    stopAlarm(true, $(this).data('snooze-time'));
+    e.preventDefault();
+  });
+
+  $('.focus-range li').click(function(e) {
+    var li = $(e.target);
+    $('.focus-range li').removeClass('selected');
+    li.addClass('selected');
+    var hours = Number(li.data('hours'));
+    foucusRangeMS = hours * 60 * 60 * 1000;
+    refreshChart();
+  });
+
+  ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+  // Client-side code to connect to server and handle incoming data
+  ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+  socket = io.connect();
+
+  function mergeDataUpdate(isDelta, cachedDataArray, receivedDataArray) {
+
+    function nsArrayDiff(oldArray, newArray) {
+      var seen = {};
+      var l = oldArray.length;
+      for (var i = 0; i < l; i++) { seen[oldArray[i].mills] = true }
+      var result = [];
+      l = newArray.length;
+      for (var j = 0; j < l; j++) { if (!seen.hasOwnProperty(newArray[j].mills)) { result.push(newArray[j]); console.log('delta data found'); } }
+      return result;
+    }
+
+    // If there was no delta data, just return the original data
+    if (!receivedDataArray) {
+      return cachedDataArray;
+    }
+
+    // If this is not a delta update, replace all data
+    if (!isDelta) {
+      return receivedDataArray;
+    }
+
+    // If this is delta, calculate the difference, merge and sort
+    var diff = nsArrayDiff(cachedDataArray,receivedDataArray);
+    return cachedDataArray.concat(diff).sort(function(a, b) {
+      return a.mills - b.mills;
+    });
+  }
+
+  socket.on('dataUpdate', function receivedSGV(d) {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [This function's cyclomatic complexity is too high. (8)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799258271/issues/source?bid=2452379&fileBranchId=2725007#l1439)